Boost: Refactor how Data Sync Entries are initialized

LiamSarsfield opened the pull request:
Fixes #39859

## Proposed changes:
Reduced `projects/plugins/boost/wp-js-data-sync.php` from **401** lines of code to ~**126** lines of code by updateing how Module specific Data Sync Entries are initialised. Instead of initialising them all within [wp-js-data-sync.php](https://github.com/Automattic/jetpack/pull/41673/files#diff-517f400908544827d36423c261ddf59bcaae2ee1df0fa7de63cf5db52ff07172), Modules can now register their own Data Sync entries within their class by applying the new `Has_Data_Sync` interface to their respective class.
The following Data Sync Module Entries have been moved:
* Image Size Analysis
* Critical CSS/Cloud CSS
* Image CDN
* Lazy Image Auto Resizer (LIAR)
* Minify CSS
* Minify JS
* Page Cache
* Performance History

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
N/A

## Does this pull request change what data or activity we track or use?
N/A

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

1. Open Chrome Dev tools, and load the Boost Settings page.
	a. Ensure there is no console errors regarding 4xx/5xx endpoints present in the logs.
2. For modules mentioned below, modify all settings for each module within Boost's Settings page.
	a. Ensure that there is no console errors regarding failed endpoints within the logs after modifying said module setting.
    b. Refresh the page, and ensure the data for said Module has persisted/saved.

* Image Size Analysis
* Critical CSS/Cloud CSS
  * Ensure that both Free Critical CSS, and Premium Critical CSS are both toggled as expected.
* Image CDN
* Lazy Image Auto Resizer (LIAR)
* Minify CSS
* Minify JS
* Page Cache
* Performance History

LiamSarsfield committed:
refactor: Update how Data Syncs are registered within Modules

LiamSarsfield committed:
changelog

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/cleanup-wp-js-data-sync

LiamSarsfield committed:
Refactor how Module Data Syncs are registered

LiamSarsfield committed:
changelog

LiamSarsfield committed:
Update data sync registration documentation and method

LiamSarsfield commented on projects/plugins/boost/app/data-sync/Performance_History_Entry.php:
```diff
@@ -1,4 +1,5 @@
 <?php
+namespace Automattic\Jetpack_Boost\Data_Sync;
```
This didn't have a namespace, not sure if that was intentional or not.

LiamSarsfield committed:
Remove unnecessary Data_Sync_Utils

LiamSarsfield committed:
Update type hint for callback parameter in Data_Sync; Remove unnecessary contructor param for Page_Cache_Entry

LiamSarsfield commented on projects/plugins/boost/app/modules/optimizations/page-cache/Page_Cache.php:
```diff
@@ -51,6 +59,39 @@ public function setup() {
+					'logging'         => Schema::as_boolean(),
+				)
+			),
+			new Page_Cache_Entry()
```
This originally had an argument passed to it with `JETPACK_BOOST_DATASYNC_NAMESPACE . '_page_cache' `, but Phan was giving out, and it doesn't appear to have a constructor, so I don't this it was used.

dilirity replied:
Yep, you're right. Nice catch.

LiamSarsfield committed:
Ensure ISA has Has_Data_Sync interface

LiamSarsfield committed:
Move Critical CSS Data Sync Setup so it can be shareable with Cloud CSS

LiamSarsfield committed:
Remove PhanParamTooMany suppression from Phan baseline

LiamSarsfield committed:
Merge branch 'trunk' into update/boost/cleanup-wp-js-data-sync

dilirity committed:
Merge branch 'trunk' into update/boost/cleanup-wp-js-data-sync

dilirity reviewed (changes requested):
It appears there's a problem with the critical css state. It's not available when this is called:

https://github.com/Automattic/jetpack/blob/7624a6e67cfbb873e91140d41b440792f4737015/projects/plugins/boost/app/lib/critical-css/Critical_CSS_State.php#L162-L164

and that generates a PHP warning:

```
Warning: Trying to access array offset on value of type null in /usr/local/src/jetpack-monorepo/projects/plugins/boost/app/lib/critical-css/Critical_CSS_State.php on line 164
```

That's because of this:

https://github.com/Automattic/jetpack/blob/c5552d189e492c2fc5e809afb03889f9ad8f2889/projects/plugins/boost/app/admin/class-admin.php#L29

which gets called before DS entries are registered.

Previously, data sync entries were loaded before the modules were registered (and the above code was executed):

https://github.com/Automattic/jetpack/blob/9ab8b9ff5cca4e760cb7c1d6da53645a66108866/projects/plugins/boost/jetpack-boost.php#L267

I think loading data sync entries here is too late:

https://github.com/Automattic/jetpack/blob/3d8358726375b8f01c04f977eaa834ad6e84014c/projects/plugins/boost/app/modules/Modules_Setup.php#L136

I can't think of a way around this

LiamSarsfield committed:
Merge branch 'trunk' into update/boost/cleanup-wp-js-data-sync

LiamSarsfield committed:
Boost: Refactor module data sync setup process

Move data sync initialization to a separate method to prevent side effects during plugin loading. This ensures data sync is set up before other modules are loaded, improving the initialization sequence.

LiamSarsfield committed:
Boost: Add method to retrieve available submodules

Refactor module initialization to use a new `get_available_submodules()` method, which simplifies the process of filtering and instantiating available submodules. This change improves code readability and reduces duplication in module setup logic.

LiamSarsfield commented:
@dilirity Thanks for noticing that.
> It appears there's a problem with the critical css state. It's not available when this is called

I've updated the code so that the Module's related data sync registries are registered immediately upon [the 
Module_Setup's setup function being called](https://github.com/Automattic/jetpack/pull/41673/files#diff-d43d5780514f1d715c5bbee67e66f52fb5d93cd10e0e03641945a5d2fb3b006bR172), instead of within the `plugins_loaded` action within `load_modules`.

It appears that the `admin_init` hook has code that relies on this Data Sync registry as you've spotted:
https://github.com/Automattic/jetpack/blob/7624a6e67cfbb873e91140d41b440792f4737015/projects/plugins/boost/app/lib/critical-css/Critical_CSS_State.php#L162-L164
But my above update should address that.

dilirity commented on projects/plugins/boost/app/contracts/Has_Data_Sync.php:
```diff
@@ -0,0 +1,15 @@
+	 *
+	 * @return void
+	 */
+	public static function register_data_sync( Data_Sync $instance );
```
Note: I noticed this is called statically, however `$feature` is an instance. Do we need it to be a static method?

LiamSarsfield replied:
Originally it was a static to ensure the helper functions also implement and follow the same registration pattern, but after refactoring from other feedback, it doesn't make sense anymore to have it as a static method so it's now an instance method.

dilirity reviewed (commented):
The problem I found during my last review is gone - nice work!

I wonder if moving registration of the data sync to `Modules_Setup` would make sense. Have `Module` "gather" information (just like we do with submodules) for it or something.

I have an itchy feeling about some features directly calling their own `Data_Sync_Setup` and others directly registering it inside `register_data_sync`. For Critical/Cloud CSS, it might be tricky to have one? Just thinking.

Here's an idea. Considering the above - wouldn't it make more sense for features to only have information about what their data sync entries are (they have been doing this so far, only containing info about what they do)? Then have the Modules_Setup initialize them using the Module instance (which gathers that information).

Tell me what you think about this idea. What we have here works, I'm just terrible at design, but I feel like the features themselves should just have a list or something and let something else handle that list.

LiamSarsfield commented:
@dilirity 

> Considering the above - wouldn't it make more sense for features to only have information about what their data sync entries are (they have been doing this so far, only containing info about what they do)? Then have the Modules_Setup initialize them using the Module instance (which gathers that information).

I do like the idea of having the Module class strictly for configuration, however from looking at how Data Sync is currently, the registration and declaration of endpoints appear to be tightly coupled. I was looking at seperating this coupling by only declaring an `Endpoint`/`Action_endpoint` to the `Module`, and having `Module_Setup` doing the actual registration, but that introduced another layer of abstraction that I think might hinder the readability of the code instead of improving it for the purpose of `Module_Setup` doing the registration instead.
https://github.com/Automattic/jetpack/blob/05b1bf265e38caabbb4989f656cceab255bc879e/projects/packages/wp-js-data-sync/src/class-registry.php#L90-L128

I will be adding a commit to this branch that I think does make it slightly more obvious what `register_data_sync` currently does, instead of delegating everything to the helper function on certain Modules. But I like the of the Module class' responsibility be only for configuration purposes, though that may involve a larger refactor of Data Sync registration that would be the primary work of the PR, instead of moving the registration outside of `wp-js-data-sync.php`

LiamSarsfield committed:
Boost: Refactor Data Sync to use instance method instead of static method

Update the `Has_Data_Sync` contract and various modules to use an instance method for registering data sync, improving consistency and flexibility in module data synchronization. This change involves:

- Modifying the `Has_Data_Sync` interface to use an instance method
- Updating multiple modules to implement the new data sync registration approach
- Removing static method calls in favor of instance method calls
- Simplifying data sync schema and registration logic in various modules

LiamSarsfield committed:
Merge remote-tracking branch 'origin/trunk' into update/boost/cleanup-wp-js-data-sync

LiamSarsfield committed:
Boost: Replace existing Module class with ModuleWrapper

Rename Module class to ModuleWrapper class to improve readability module management in Jetpack Boost:
- Replace existing Module class with ModuleWrapper

LiamSarsfield commented on projects/plugins/boost/app/modules/ModuleWrapper.php:
I've renamed the `Module` class to `ModuleWrapper` here mainly to improve readability. The `Module`/`ModuleWrapper `class' primary responsibility is to interact with the `feature` (now `module`) property instead of being the actual Module itself.

LiamSarsfield committed:
Boost: Refactor Modules Index and Setup to Improve Module Management

- Add `get_modules()` method to `Modules_Index` to retrieve all modules
- Update `Modules_Setup` to use `get_modules()` for data sync setup
- Remove unused data sync registration in `Cloud_CSS`
- Modify module initialization to store all modules, not just available ones

LiamSarsfield committed:
Boost: Remove Has_Data_Sync interface from Cloud_CSS module

Remove unnecessary data sync interface from Cloud_CSS module, following previous refactoring of module data synchronization

LiamSarsfield committed:
Boost: Update Phan Baseline for Modules and Minify Files

Update Phan baseline configuration to address new type checking warnings in Boost modules:
- Add new Phan issue types for undeclared methods, properties, and types
- Update baseline for Modules_Index and Modules_Setup
- Adjust type checking suppressions for minify and CLI-related files

dilirity commented on projects/plugins/boost/app/modules/ModuleWrapper.php:
![CleanShot 2025-02-19 at 15 08 23](https://github.com/user-attachments/assets/ea50de32-f654-4b46-a382-0c086f7dc3ea)

ModuleWrapper is a [pretender](https://www.youtube.com/watch?v=SBjQ9tuuTJQ).

Use snake_case instead of PascalCase.

LiamSarsfield replied:
Thanks for catching this! Since I reverted the `ModuleWrapper` change this isn't needed.

dilirity commented on projects/plugins/boost/app/modules/Modules_Index.php:
```diff
@@ -47,8 +53,9 @@ class Modules_Index {
 	 */
 	public function __construct() {
 		foreach ( self::FEATURES as $feature ) {
+			$this->modules[ $feature::get_slug() ] = new ModuleWrapper( new $feature() );
```
Note: Doing this will always initialize the contructor of a feature. Mentioning this in case we have some features that are doing things they shouldn't be.

I looked through our features and couldn't find anything that wouldn't have happened before. For example, Page_Cache reads a file on every page load. Previously that happened only when the feature was available. However, since it's always available (it's free 🥳), this is not a problem.

LiamSarsfield replied:
Makes sense, as you said, they shouldn't be 😅 but could be. Thankfully not though.

dilirity commented on projects/plugins/boost/app/modules/ModuleWrapper.php:
```diff
@@ -0,0 +1,90 @@
+	/**
+	 * @var Pluggable
+	 */
+	public $module;
```
Can we not rename this to `module`?

To me, it makes sense for a feature to be a part of a module, since a module can contain multiple features (in our case one).

We already have `feature` as an established term throughout the codebase.

LiamSarsfield replied:
Per offline discussion with @dilirity, it may be worthwhile to investigate if we can reduce confusion between a module and a feature, however due to the potential refactor effort this will cause, I've split this out and reverted this renaming.
https://github.com/Automattic/jetpack/issues/41891

dilirity commented on projects/plugins/boost/app/modules/Modules_Setup.php:
```diff
@@ -77,6 +81,30 @@ public function register_always_available_endpoints( $feature ) {
 
+	public function setup_modules_data_sync( $modules ) {
+		foreach ( $modules as $module ) {
+			$this->register_module_data_sync( $module->module );
```
`$module->module` :D

LiamSarsfield replied:
Yeah this is code smelly 🫠

dilirity commented on projects/plugins/boost/app/modules/Modules_Setup.php:
```diff
@@ -77,6 +81,30 @@ public function register_always_available_endpoints( $feature ) {
+	 *
+	 * @return bool|void
+	 */
+	public function register_module_data_sync( $feature ) {
```
I think this needs to be `$module`? I think this might be the reason I was advocating for the module handling the registration instead of the feature. Maybe this can go in `Module_Wrapper`? Uh it probably can't because we also have data sync in `Modules_Setup`... Our code base/naming is such a mess.

LiamSarsfield replied:
Good point, I've updated the function to be `register_feature_data_sync` instead of module specific.
I do agree that this function should be only fetching the data sync and then registering the entries/data sync here. Though I think that may be a separate effort from this, the data sync's registration and action declaration is [currently tightly coupled](https://github.com/Automattic/jetpack/pull/41673#issuecomment-2665883152).

dilirity commented on projects/plugins/boost/app/modules/image-size-analysis/Image_Size_Analysis.php:
```diff
@@ -2,16 +2,33 @@
 		Image_Size_Analysis_Fixer::setup();
 	}
 
+	public function register_data_sync( Data_Sync $instance ) {
+		$instance->register( 'image_size_analysis', Data_Sync_Schema::image_size_analysis(), new Image_Size_Analysis_Entry() );
+		$instance->register_action( 'image_size_analysis', 'paginate', Data_Sync_Schema::image_size_analysis_paginate(), new Image_Size_Analysis_Summary_Action_Paginate() );
+		$instance->register_action( 'image_size_analysis', 'fix', Data_Sync_Schema::image_size_analysis_fix(), new Image_Analysis_Action_Fix() );
+		$instance->register( 'image_size_analysis_summary', Data_Sync_Schema::image_size_analysis_summary(), new Image_Size_Analysis_Summary() );
+		$instance->register_action( 'image_size_analysis_summary', 'start', Schema::as_void(), new Image_Size_Analysis_Summary_Action_Start() );
+	}
+
```
I like this way better!

dilirity commented on projects/plugins/boost/app/modules/optimizations/page-cache/Page_Cache.php:
```diff
@@ -51,6 +59,34 @@ public function setup() {
 	}
 
+	public function register_data_sync( Data_Sync $instance ) {
+		$page_cache_schema       = Schema::as_assoc_array(
+			array(
+				'bypass_patterns' => Schema::as_array( Schema::as_string() ),
+				'logging'         => Schema::as_boolean(),
+			)
+		);
+		$page_cache_error_schema = Schema::as_assoc_array(
+			array(
+				'code'      => Schema::as_string(),
+				'message'   => Schema::as_string(),
+				'dismissed' => Schema::as_boolean()->fallback( false ),
+			)
+		)->nullable();
```
How do we decide when to move schemas to a separate class?

LiamSarsfield replied:
Good question, I personally don't have a strict criteria on when or if we should move schemas. I think if the schemas take up the largest portion of the class, and/or they make it difficult to see what is being registered; they should be moved. But let me know if you think otherwise

dilirity replied:
Yeah, I think what you're saying makes sense. We can't have a strict rule about this. If it hinders readability - move it. Otherwise, it's fine to be in the main feature class.

dilirity reviewed (changes requested):
Awesome!

I think we're good to go, just a few minor things and we can merge.

LiamSarsfield committed:
Boost: Rename ModuleWrapper to Module and Refactor Module Management

- Replace ModuleWrapper class with Module class
- Update references to ModuleWrapper across Boost plugin files
- Modify terminology from "module" to "feature" in comments and method names
- Simplify module initialization and data sync registration
- Improve code readability and consistency in module handling

LiamSarsfield committed:
Boost: Update Phan Baseline to Reduce Type Checking Warnings

LiamSarsfield committed:
Merge branch 'trunk' into update/boost/cleanup-wp-js-data-sync

dilirity commented on projects/plugins/boost/app/modules/optimizations/critical-css/Critical_CSS.php:
```diff
@@ -69,6 +78,17 @@ public function setup() {
 	}
 
+	public function register_data_sync( Data_Sync $instance ) {
+		// TODO: Refactor to remove repition with Cloud CSS.
```
TODO left in here. Does it need to be addressed here or in a follow-up?

LiamSarsfield replied:
Good catch, this TODO is addressed. Will remove it now.

dilirity reviewed (dismissed):
I think we can merge this!

There was a stray todo I've noted below.

LiamSarsfield committed:
Remove TODO comment that is done

dilirity reviewed (approved):
:shipit:
